### PR TITLE
Add implementation of WriterT and WriterTMonad

### LINKED ISSUE #84: Add WriterT monad
Taking this one because I've never looked into the writer monad so, yeah.

### PR DESCRIPTION
**Pull request by pakoito**
Closes #84

### CODE REVIEW

**Comment by pakoito** [comment]
Missing: tests

**Comment by pakoito** [comment]
I went overboard and fixed the Monoid and Semigroup instances.

File: katz/src/main/kotlin/katz/instances/NonEmptyListSemigroup.kt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package katz
+
+class NonEmptyListSemigroup<A> : Semigroup<NonEmptyList<A>> {
+    override fun combine(a: NonEmptyList<A>, b: NonEmptyList<A>): NonEmptyList<A> =
+            NonEmptyList.fromListUnsafe(a.all + b.all)
```
Comment (jrgonzalezg) [inline katz/src/main/kotlin/katz/instances/NonEmptyListSemigroup.kt]: This is the same as just `a + b`

Comment (pakoito) [inline katz/src/main/kotlin/katz/instances/NonEmptyListSemigroup.kt]: No, it won't pick it up :D

Comment (jrgonzalezg) [inline katz/src/main/kotlin/katz/instances/NonEmptyListSemigroup.kt]: Committed a fix :)

File: katz/src/main/kotlin/katz/typeclasses/Applicative.kt [hunk 2]
```diff
@@ -20,7 +20,9 @@ interface Applicative<F> : Functor<F>, Typeclass {
             fb.map { fb -> map2(fa, fb, f) }
 }
 
-data class Tuple2<out A, out B>(val a: A, val b: B)
+data class Tuple2<out A, out B>(val a: A, val b: B) {
+    fun swap(): Tuple2<B, A> = Tuple2(b, a)
```
Comment (raulraja) [inline katz/src/main/kotlin/katz/typeclasses/Applicative.kt]: what if we call this `reverse` and implement it in all Tuple*?

File: katz/src/main/kotlin/katz/typeclasses/Applicative.kt [hunk 3]
```diff
@@ -30,6 +32,10 @@ data class Tuple8<out A, out B, out C, out D, out E, out F, out G, out H>(val a:
 data class Tuple9<out A, out B, out C, out D, out E, out F, out G, out H, out I>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F, val g: G, val h: H, val i: I)
 data class Tuple10<out A, out B, out C, out D, out E, out F, out G, out H, out I, out J>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F, val g: G, val h: H, val i: I, val j: J)
 
+infix fun <A, B, C, D> Tuple3<A, B, C>.toT(d: D): Tuple4<A, B, C, D> = Tuple4(this.a, this.b, this.c, d)
```
Comment (raulraja) [inline katz/src/main/kotlin/katz/typeclasses/Applicative.kt]: Should this be `plus`? Also wondering what a `Semigroup` instance for each of the tuples looks like ;)

**Pull request merged**